### DISCUSSION TITLE
How to define filter for the nested JSON array?

### DISCUSSION THREAD

**Discussion by kirdk** [mikro-orm/mikro-orm#3718] [Show and tell]
Hello everyone,
Can anyone clarify how to filter entities with nested the array stored in JSON or string DB field?
For instance, we have an entity Program:
```
@Entity()
export class Program {
  @PrimaryKey()
  id: string = uuidv4();

  @Property({ type: JsonType })
  systems: string[];
}
```
How should I define the filter to find all programs which contain 'Windows'? 
How should I define the filter to find all programs which contain only 'Windows'? 

I've tried to define it as `this.em.findAndCount(ent.Program, { systems: 'WINDOWS' } }`  and `systems: ['WINDOWS']` but it seems both are not working for the JSON arrays.

**Comment by B4nan** [comment]
What SQL query do you want to produce?

**Reply by kirdk** [reply]
I'm trying to build an API and use ORM queries as options sent from the client (something like /api/program?filter=systems: 'Windows'). I'm just trying to understand how it works for different cases. It works fine for the scalar fields and nested objects/collections. But it works unexpectedly for the nested arrays. 
So, in my opinion, it would be good to have `$contains` operator for all drivers, which can be converted to `like` SQL expression

**Reply by B4nan** [reply]
I asked for SQL query. What SQL query do you expect from this?

Note that `$contains` is a postgres array operator, it does not have much in common with `$like`, both sides for `$contains` need to be array and it checks whether the array on the right-hand side is contained in the one on the left-hand side. You cannot use `$contains` for querying a JSON column at all.

**Reply by kirdk** [reply]
To find all Programs where 'Windows' is contained in the array I think I should have `where systems like %Windows%`

**Reply by B4nan** [reply]
If you want like query, use `$like` operator.

**Reply by nmichkarev** [reply]
For example above, if we have an ArrayType for systems property and try to query `this.em.findAndCount(ent.Program, { systems: ['WINDOWS'] } }`, it creates sql with condition `where systems in ('WINDOWS')`. Is it a proper behaviour? Maybe it would be better if orm created something like `where systems = '{WINDOWS}'`?

If would also be nice to have a query operator to match an array that contains element for ArrayType.

**Reply by B4nan** [reply]
> For example above, if we have an ArrayType

But `ArrayType` is not the same as `JsonType`, those are very different things. And the OP is asking about using JSON properties.

> Is it a proper behaviour?

Yes, it is expected behaviour, arrays on right hand side are always considered as IN queries. This is [documented](https://mikro-orm.io/docs/query-conditions).

Maybe we could improve this for array properties - but this issue was about JSON properties and that's a different story.

> If would also be nice to have a query operator to match an array that contains element for ArrayType.

It is there, for postgres. The ORM operators are mapping SQL operators, I can't add new ones when I don't have any SQL feature I would map to. And that is why I am asking from the very beginning about what SQL queries people expect.

> Maybe it would be better if orm created something like where systems = '{WINDOWS}'?

I see you are actually using postgres, given the use of `{}` in the array literal (next time at least mention it, we are talking about driver-specific features, and I also still have no idea what driver is the OP about). Postgres have full support for this, but you need to use the operators, based on what you want. If you want to compare the values, use `$eq`, or if you want one of the array operators: 

<img width="415" alt="image" src="https://user-images.githubusercontent.com/615580/201033665-12685175-227f-484e-b6c7-a615f7b52765.png">

https://mikro-orm.io/docs/query-conditions

---

So once again, let's start taking details, I really want you guys to provide valid SQL for what you want. Then we can talk how you can produce it with the ORM, or whether we should adjust some of the existing behaviors.

**Reply by kirdk** [reply]
Dear @B4nan, it's not so important whether to use JSON array or ArrayType, the aim is to have the possibility to store a nested array without additional entity. It also is not a topic about how to build specific SQL, because if we understand how the query builder is working we can build any query. Our question is about the concept and about how to make it maybe a bit more generic and logical.
In our opinion, it's strange that an array should be filtered as a string (with ArrayType). Because someone from the team can work only with the object returned by ORM and there it is an array, they can't know how is it stored. 
And another strange thing we found is that $like operator works in the same way as $eq and we should add % manually...

**Reply by B4nan** [reply]
> Dear @B4nan, it's not so important whether to use JSON array or ArrayType,

It is very important, as I said, internally those are two very different things. You clearly don't see the difference, that doesn't mean it is not there.

> In our opinion, it's strange that an array should be filtered as a string (with ArrayType).

I haven't said anything like that 🤷 Only thing I said is that if you want to have a `like` operator in your query, you need to use it yourself and don't expect the ORM to do it for you.

> And another strange thing we found is that $like operator works in the same way as $eq and we should add % manually...

Now that's a very weird expectation. I don't think there is a single tool doing that.
